gobblin-core/src/main/java/gobblin/publisher/TimePartitionedDataWithRowCountsPublisher.java

```diff
@@ -0,0 +1,160 @@
+import gobblin.configuration.State;
+import gobblin.configuration.WorkUnitState;
+import gobblin.util.ForkOperatorUtils;
+import gobblin.util.WriterUtils;
+import gobblin.writer.WriterOutputFormat;
+import java.io.IOException;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Set;
+import org.apache.hadoop.fs.FileStatus;
+import org.apache.hadoop.fs.Path;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * For time partition jobs, writer output directory is
+ * $GOBBLIN_WORK_DIR/task-output/{extractId}/{tableName}/{partitionPath},
+ * where partition path is the time bucket, e.g., 2015/04/08/15.
+ *
+ * While moving the file, also places the row count (number of entries contained within the file)
+ * into the name of the file such that output.avro becomes output.{rowCount}.avro.
+ *
+ * Unlike {@link TimePartitionedDataPublisher}, moves each file individually rather than moving
+ * directories at a time. This is done so that each task/branch can move the file it created,
+ * allowing it to access the row count information and place it into the file name.
+ *
+ * Publisher output directory is $GOBBLIN_WORK_DIR/job-output/{tableName}/{partitionPath}
+ *
+ * @author ekrogen
+ */
+public class TimePartitionedDataWithRowCountsPublisher extends TimePartitionedDataPublisher {
+
+  private static final Logger LOG = LoggerFactory.getLogger(TimePartitionedDataWithRowCountsPublisher.class);
+
+  public TimePartitionedDataWithRowCountsPublisher(State state) {
+    super(state);
+  }
+
+  @Override
+  public void publishData(Collection<? extends WorkUnitState> states) throws IOException {
+
+    Set<Path> publisherDirsProcessed = new HashSet<Path>();
+
+    for (WorkUnitState workUnitState : states) {
+      for (int branchId = 0; branchId < this.numBranches; branchId++) {
+
+        // The directory where the workUnitState wrote its output data. It is a combination of
+        // WRITER_OUTPUT_DIR and WRITER_FILE_PATH
+        Path writerOutputDir = WriterUtils.getWriterOutputDir(workUnitState, this.numBranches, branchId);
+
+        if (!this.fss.get(branchId).exists(writerOutputDir)) {
+          LOG.warn("WorkUnit " + workUnitState.getId() + " produced no data");
+          workUnitState.setWorkingState(WorkUnitState.WorkingState.COMMITTED);
+          break;
+        }
+
+        String fileExtension = WriterOutputFormat.valueOf(workUnitState.getProp(ForkOperatorUtils
+                .getPropertyNameForBranch(ConfigurationKeys.WRITER_OUTPUT_FORMAT_KEY, this.numBranches, branchId),
+            WriterOutputFormat.AVRO.name())).getExtension();
+
+        String writerOutputFileName = WriterUtils.getWriterFileName(workUnitState, this.numBranches,
+            branchId, workUnitState.getId(), fileExtension);
+
+        Path writerOutputFullGlob =
+            getWriterOutputGlobMatcher(workUnitState, branchId, writerOutputDir, writerOutputFileName);
+
+        Path writerOutputFilePath = null;
```

Comment:
Use `com.google.common.base.Optional` instead of null: https://code.google.com/p/guava-libraries/wiki/UsingAndAvoidingNullExplained

(don't use `java.util.Optional` since we need to support Java 6 & 7).